src/darwin/Framework/CHIP/MTRMetricsCollector.mm

```diff
@@ -0,0 +1,203 @@
+/**
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#import "MTRMetricsCollector.h"
+#import "MTRLogging_Internal.h"
+#include "MTRMetrics_Internal.h"
+#import <MTRUnfairLock.h>
+#include <platform/Darwin/Tracing.h>
+#include <tracing/metric_event.h>
+#include <tracing/registry.h>
+
+using MetricEvent = chip::Tracing::MetricEvent;
+
+static NSString * kMTRMetricDataValueKey = @"value";
+static NSString * kMTRMetricDataTimepointKey = @"time_point";
+static NSString * kMTRMetricDataDurationKey = @"duration_us";
+
+void InitializeMetricsCollection()
+{
+    if ([MTRMetricsCollector sharedInstance]) {
+        MTR_LOG_INFO("Initialized metrics collection backend for Darwin");
+    }
+}
+
+@implementation MTRMetricsData {
+    chip::System::Clock::Microseconds64 _timePoint;
+    chip::System::Clock::Microseconds64 _duration;
+}
+
+- (instancetype)initWithMetricEvent:(const MetricEvent &)event
+{
+    if (!(self = [super init])) {
+        return nil;
+    }
+
+    if (event.value.type == MetricEvent::Value::Type::Signed32Type) {
+        _value = [NSNumber numberWithInteger:event.value.store.int32_value];
+    } else {
+        _value = [NSNumber numberWithUnsignedInteger:event.value.store.uint32_value];
+    }
+    _timePoint = event.timePoint;
+    _duration = chip::System::Clock::Microseconds64(0);
+    return self;
+}
+
+- (void)setDurationFromMetricData:(MTRMetricsData *)fromData
+{
+    _duration = _timePoint - fromData->_timePoint;
+}
+
+- (NSNumber *)timePointMicroseconds
+{
+    return [NSNumber numberWithUnsignedLongLong:_timePoint.count()];
+}
+
+- (NSNumber *)durationMicroseconds
+{
+    return [NSNumber numberWithUnsignedLongLong:_duration.count()];
+}
+
+- (NSString *)description
+{
+    return [NSString stringWithFormat:@"Value = %@, TimePoint = %@, Duration = %@ us", self.value, self.timePointMicroseconds, self.durationMicroseconds];
```

Comment:
Should this not include the string MTRMetricsData somewhere?

Reply:
Done!